### PR TITLE
Aggregation v2

### PR DESCRIPTION
**Pull request by Riateche**
- Include oracle program as a dependency
- Move all pyth functions from `bank.rs` to a submodule and remove unnecessary arguments
- Aggregate prices and create price feed messages for price feeds with v2 aggregation enabled
- Merge v1 and v2 messages and generate message accounts and wormhole accounts for all of them

### CODE REVIEW

File: runtime/src/bank/pyth_accumulator.rs [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+        ),
+        (
+            "ACCUMULATOR_SEQUENCE_ADDR",
+            Pubkey::new_from_array(pythnet::ACCUMULATOR_SEQUENCE_ADDR),
+        ),
+        (
+            "WORMHOLE_PID",
+            Pubkey::new_from_array(pythnet::WORMHOLE_PID),
+        ),
+    ];
```
Comment (ali-behjati) [inline runtime/src/bank/pyth_accumulator.rs]: add oracle pubkey here too.

File: runtime/src/bank/pyth_accumulator.rs [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+            value
+                .parse()
+                .expect("invalid value of PYTH_ORACLE_PUBKEY env var")
+        ),
+        Err(VarError::NotPresent) => None,
+        Err(VarError::NotUnicode(err)) => {
+            panic!("invalid value of PYTH_ORACLE_PUBKEY env var: {err:?}");
+        }
+    };
+}
```
Comment (ali-behjati) [inline runtime/src/bank/pyth_accumulator.rs]: I suggest making it consistent with other env vars (or changing them to be like this). Also the name `ORACLE_PID` matches the other env vars more.

File: runtime/src/bank/pyth_accumulator.rs [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    acc: pythnet_sdk::accumulators::merkle::MerkleAccumulator<
+        pythnet_sdk::hashers::keccak256_160::Keccak160,
+    >,
+    ring_index: u32,
+) -> std::result::Result<(), AccumulatorUpdateErrorV1> {
+    use pythnet_sdk::{
+        pythnet,
+        wormhole::{AccumulatorSequenceTracker, MessageData, PostedMessageUnreliableData},
+        ACCUMULATOR_EMITTER_ADDRESS,
+    };
```
Comment (ali-behjati) [inline runtime/src/bank/pyth_accumulator.rs]: move these up?

**Review by ali-behjati — COMMENTED** [review]
Nice.

File: runtime/src/bank/pyth_accumulator.rs (hunk 1 above)
Comment (Riateche) [reply runtime/src/bank/pyth_accumulator.rs]: Done.

File: runtime/src/bank/pyth_accumulator.rs (hunk 2 above)
Comment (Riateche) [reply runtime/src/bank/pyth_accumulator.rs]: Renamed and unified with other vars.

File: runtime/src/bank/pyth_accumulator.rs (hunk 3 above)
Comment (Riateche) [reply runtime/src/bank/pyth_accumulator.rs]: Done.

File: runtime/src/bank/pyth_accumulator.rs [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+        "ACCUMULATOR_SEQUENCE_ADDR",
+        Pubkey::new_from_array(pythnet_sdk::pythnet::ACCUMULATOR_SEQUENCE_ADDR),
+    );
+    pub static ref WORMHOLE_PID: Pubkey = env_pubkey_or(
+        "WORMHOLE_PID",
+        Pubkey::new_from_array(pythnet_sdk::pythnet::WORMHOLE_PID),
+    );
+    pub static ref ORACLE_PID: Pubkey = env_pubkey_or(
+        "ORACLE_PID",
+        "FsJ3A3u2vn5cTVofAjvy6y5kwABJAqYWpe4975bi2epH"
```
Comment (ali-behjati) [inline runtime/src/bank/pyth_accumulator.rs:42]: not for now, but seeing this i'm wondering whether in the future it makes sense to move the entire code in this file to pythnet_sdk

File: runtime/src/bank/pyth_accumulator_tests.rs [hunk 5]
```diff
@@ -0,0 +1,904 @@
...
+                    .as_twap_message(&price_feed_key.to_bytes().into())
+                    .to_bytes(),
+            ]
+        };
+
+        bank.store_account(&price_feed_key, &price_feed_account);
+
+        if generate_buffers {
+            // Insert into message buffer in reverse order to test that accumulator
+            // sorts first.
```
Comment (ali-behjati) [inline runtime/src/bank/pyth_accumulator_tests.rs]: where are you reversing it :?

File: runtime/src/bank/pyth_accumulator_tests.rs [hunk 6]
```diff
@@ -0,0 +1,904 @@
...
+    // 1. Create Price Feed Accounts owned by ORACLE_PUBKEY
+    // 2. Populate Price Feed Accounts
+    // 3. Call update_v2()
+    //    - Cases:
+    //      - No V1 Messages, Only Price Accounts with no V2
+    //      - No V1 Messages, Some Price Accounts with no V2
+    //      - Some V1 Messages, No Price Accounts with no V2
+    //      - Some V1 Messages, Some Price Accounts with no V2
+    //      - Simulate PriceUpdate that WOULD trigger a real V1 aggregate before End of Slot
+    //      - Simulate PriceUpdate that doesn't trigger a real V1 aggregate, only V2.
```
Comment (ali-behjati) [inline runtime/src/bank/pyth_accumulator_tests.rs]: i think this is outdated :?

**Review by ali-behjati — APPROVED** [review]
Nice!

File: runtime/src/bank/pyth_accumulator_tests.rs (hunk 6 above)
Comment (Riateche) [reply runtime/src/bank/pyth_accumulator_tests.rs]: Yes. Removed.

File: runtime/src/bank/pyth_accumulator_tests.rs (hunk 5 above)
Comment (Riateche) [reply runtime/src/bank/pyth_accumulator_tests.rs]: This was left over after copy-paste. Removed.

File: runtime/src/bank/pyth_accumulator.rs [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    vec![
+        ("MESSAGE_BUFFER_PID", Ok(*MESSAGE_BUFFER_PID)),
+        ("ACCUMULATOR_EMITTER_ADDR", Ok(*ACCUMULATOR_EMITTER_ADDR)),
+        ("ACCUMULATOR_SEQUENCE_ADDR", Ok(*ACCUMULATOR_SEQUENCE_ADDR)),
+        ("WORMHOLE_PID", Ok(*WORMHOLE_PID)),
+        ("ORACLE_PID", Ok(*ORACLE_PID)),
+    ]
+}
+
+pub fn update_v1<'a>(
```
Comment (guibescos) [inline runtime/src/bank/pyth_accumulator.rs]: lifetime seems unnecessary

File: runtime/src/bank/pyth_accumulator_tests.rs [hunk 8]
```diff
@@ -0,0 +1,890 @@
...
+
+    // Get the wormhole message generated by freezed. We don't need
+    // to offset the ring index as our test is always below 10K slots.
+    let wormhole_message_account = get_wormhole_message_account(&bank, bank.slot() as u32);
+    assert_ne!(wormhole_message_account.data().len(), 0);
+    let deserialized_wormhole_message =
+        PostedMessageUnreliableData::deserialize(&mut wormhole_message_account.data()).unwrap();
+
+    // Create MerkleAccumulator by hand to verify that the Wormhole message
+    // contents are correctg.
```
Comment (guibescos) [inline runtime/src/bank/pyth_accumulator_tests.rs]: typo: correctg

File: runtime/src/bank/pyth_accumulator_tests.rs [hunk 9]
```diff
@@ -0,0 +1,890 @@
...
+
+    let messages = prices_with_messages
+        .iter()
+        .map(|(_, messages)| messages)
+        .flatten()
+        .map(|message| &message[..])
+        .sorted_unstable()
+        .dedup()
+        .collect::<Vec<_>>();
+
```
Comment (guibescos) [inline runtime/src/bank/pyth_accumulator_tests.rs:782]: nitpick: I'd assert the length of `messages` here, to confirm they haven't been deduplicated

**Review by guibescos — APPROVED** [review]

File: runtime/src/bank/pyth_accumulator.rs (hunk 7 above)
Comment (Riateche) [reply runtime/src/bank/pyth_accumulator.rs]: Fixed.

File: runtime/src/bank/pyth_accumulator_tests.rs (hunk 8 above)
Comment (Riateche) [reply runtime/src/bank/pyth_accumulator_tests.rs]: Fixed.

File: runtime/src/bank/pyth_accumulator_tests.rs (hunk 9 above)
Comment (Riateche) [reply runtime/src/bank/pyth_accumulator_tests.rs:782]: Added.

**Review by ali-behjati — APPROVED** [review]

**Pull request merged**